winsup/cygwin/path.cc

```diff
@@ -2220,6 +2311,77 @@ symlink_worker (const char *oldpath, path_conv &win32_newpath, bool isdevice)
 	}
       else /* wsym_type == WSYM_sysfile */
 	{
+          if (wsym_type == WSYM_deepcopy)
+	    {
+	      path_conv src_path;
+	      src_path.check (oldpath, PC_SYM_NOFOLLOW, stat_suffixes);
+	      if (src_path.error)
+		{
+		  set_errno (src_path.error);
+		  __leave;
+		}
+	      if (!src_path.isondisk () && !src_path.is_fs_special ())
```

Comment:
I don't think that's right... perhaps
```suggestion
	      if (src_path.isondisk ())
```

Reply:
I don't know for sure, I was trying to reason about the original, and then started trying to reason about what you changed it to... so I'm all sorts of mixed up on what it's trying to do atm :grin:

Reply:
the original was
```c++
if (!src_path.isdevice () && !src_path.is_fs_special ())
```

isdevice was
```c++
return dev.not_device (FH_FS) && dev.not_device (FH_FIFO);
```
where not_device is
```c++
return d.devn && n != d.devn;
```

is_fs_special is
```c++
return dev_on_fs && d.devn != FH_FS;
```


so expanded that would be
```c++
if ((!d.devn || d.devn == FH_FS || !d.devn || d.devn == FH_FIFO) && (!dev_on_fs || d.devn == FH_FS))
```

the new isondisk is
```c++
dev_on_fs || d.devn == FH_FS
```

so I'm pretty sure `!isondisk ()` is roughly equivalent, with the possible exception of devn == 0 or devn == FH_FIFO.

Reply:
Actually, `(A || B) && A` reduces to `A` (where `A` is `!d.devn || d.devn == FH_FS` and `B` is `!d.devn || d.devn == FH_FIFO`). The `B` does not matter. It's a wasteful (and misleading) addition to the condition.

Therefore, the previous `!src_path.isdevice () && !src_path.is_fs_special ()` and the new `src_path.isondisk ()` are not only roughly identical, they are exactly identical (modulo performance if the compiler does not optimize the useless condition out), unless `dev_on_fs` does something that I missed.

Reply:
I don't know, I'm having a hard time reasoning about this, and I don't have a lot of attention to give this today, but the `dev_on_fs` vs `!dev_on_fs` is throwing me.

If it's `FH_FS` it doesn't matter.  The only other cases where it might is `devn == 0` or `FH_FIFO`, in which case `!dev_on_fs` would have had to have been true, now if it's not `FH_FS` `dev_on_fs` would have to be true instead.  I almost want to say `!src_path.isspecial()` might be what was intended here. But this is an odd edgecase (trying to figure out what to do if somebody tried to do a deepcopy symlink of a special file)

Reply:
Oh darn it, forget my `(A || B) && A` reasoning: I missed that the first `A` starts with `!d.devn` and the second `A` starts with `!dev_on_fs`, i.e. the second `A` is actually a `C` and I was wrong...

Reply:
Okay, let's try again:

```
(!d.devn || d.devn == FH_FS || !d.devn || d.devn == FH_FIFO) && (!dev_on_fs || d.devn == FH_FS)
```

Leaving out the duplicate `!d.devn`:

```
(!d.devn || d.devn == FH_FS || d.devn == FH_FIFO) && (!dev_on_fs || d.devn == FH_FS)
```

With `&&` binding stronger than `||` we can expand to:

```
!d.devn && !dev_on_fs || d.devn == FH_FS && !dev_on_fs || d.devn == FH_FIFO && !dev_on_fs || !d.devn && d.devn == FH_FS || d.devn == FH_FS && d.devn == FH_FS || d.devn == FH_FIFO && d.devn == FH_FS
```

Leaving out the impossible (like `!d.devn && d.devn  == FH_FS`):

```
!d.devn && !dev_on_fs || d.devn == FH_FS && !dev_on_fs || d.devn == FH_FIFO && !dev_on_fs || d.devn == FH_FS && d.devn == FH_FS
```

No, we can simplify `d.devn == FH_FS && d.devn == FH_FS` to `d.devn == FH_FS` and since that condition strictly encloses any condition `d.devn == FH_FS && ...` we can also drop the latter, i.e.

```
!d.devn && !dev_on_fs || d.devn == FH_FIFO && !dev_on_fs || d.devn == FH_FS
```

Since `dev_on_fs` can only be non-zero if `setfs(1)` is called, and in all three code-paths where that happens, the `devn` is parsed to a non-zero value first, `!d.devn` already implies `!dev_on_fs` and we can simplify the first term:

```
!d.devn || d.devn == FH_FIFO && !dev_on_fs || d.devn == FH_FS
```

Now, if `d.devn == FH_FIFO && !dev_on_fs` holds, then `d.devn != FH_FS` holds, too, (because it is implied) and therefore we can rewrite this condition as `d.devn == FH_FIFO && !(dev_on_fs || d.devn == FH_FS)` i.e. it is equivalent to `src_path.isfifo() && !src_path.isondisk()`

The `!d.devn || d.devn == FH_FS` condition can be rewritten as `!(d.devn && d.devn != FH_FS)`, which is equivalent to `!dev.not_device(FH_FS)`, i.e. `!src_path.isspecial()`.

So we can rewrite the original condition as `!src_path.isspecial() || (src_path.isfifo() && !src_path.isondisk())`.

However, that `src_path.isfifo() && !src_path.isondisk()` part looks suspicious. The code we're trying to adjust is the code path where the MSYS2 runtime is asked to make a symlink, and it makes a deep copy instead. Why would it make a deep copy of a FIFO, but only if it is not on disk (and would `CopyFile()` handle that at all, it would probably fail)? That part of the condition seems bogus.

Now, as the saying goes "If you want to know whether you truly understand something, try to explain it" and as the original code comment "MSYS copy file instead make symlink" explains precisely nothing at all, there is a very much non-zero chance that the original author did not understand the logic.

As a consequence, I maintain that the original condition was incorrect from the start. It would be not surprising at all, given that the entire idea of letting `symlink()` "succeed" by doing a deep copy is highly dubious to begin with, and I would love to have the time to push forward #114 further (which would require a deep dive into the question why Git's test suite fails with that).

In the meantime, and to unblock this PR, I would like to circle back to the question what condition _would_ be correct. We want to do a deep copy, and naturally we cannot copy certain entities (such as disk devices), and we would want to fail the operation in that case. I could imagine that the correct thing would be to drop the condition altogether: If a symlink should be emulated via a deep copy, and if that involves an entity that cannot be copied (because it is not on disk), then the operation should fail (instead of skipping the attempt to copy said entity).

Alternatively, I would suggest that we either go with the `!src_path.isspecial()` condition (which would align better with the original condition than `src_path.isondisk()`, and I just [verified](https://github.com/git-for-windows/msys2-runtime/actions/runs/12989859693?pr=84) that it at least does not break Git's test suite), or with `src_path_isondisk()`.

I would actually much prefer `src_path.isondisk()` because that would be a rare case of a readable and intuitively understandable construct in the code base.

@jeremyd2019 if you agree, I will reduce the `fixup!` commits into a single one so that I have a place where to jot down the reasoning in a commit message.

Reply:
My understanding is that if that condition is false, it will instead fall back to making a `WSYM_sysfile` style symlink, not fail.  Windows apps (like Git for Windows) wouldn't know what to do with those, but they wouldn't know what to do with a Cygwin 'special' file either, so I expect that's ok.

I think `!src_path.isspecial()` because of that.

If there's a time and place to investigate/improve the deepcopy code (#114? or somewhere else), that looks like a good idea...  It only looks at the direct src_path to see if it's special, but then `recursiveCopy` just uses win32 functions, so it would have no idea if an on-disk special file was in there.  (also, I think it's using the ANSI Windows functions, plus is limited to `MAX_PATH`).

Reply:
yeah, try
```bash
mknod foo c 1 3
ln -s foo foolnk
```
see in Windows you've got `foo.lnk` for the device file (see the isdevice check at the beginning of the function, that comes from mknod not the former pathconv.isdevice()), and foolnk which is a sysfile-style symlink.

Reply:
> ... `setfs(1)` is called, and in all three code-paths where that happens ...

BTW, this is back down to two code-paths after https://github.com/cygwin/cygwin/commit/5a58110d12842a477c4ff9f3f2df1c0175b134ce

Reply:
> My understanding is that if that condition is false, it will instead fall back to making a `WSYM_sysfile` style symlink, not fail.

Ah, I missed that, thank you for pointing that out!

> try
> 
> ```
> $ mknod foo c 1 3
> $ ln -s foo foolnk
> $ ls -l foo*
> crw-rw-rw- 1 x None 1, 3 Jan 27 10:43 foo
> lrwxrwxrwx 1 x None    3 Jan 27 10:43 foolnk -> foo
> ```
> 
> see in Windows you've got `foo.lnk` for the device file (see the isdevice check at the beginning of the function, that comes from mknod not the former pathconv.isdevice()), and foolnk which is a sysfile-style symlink.

Excellent! I turned this into the following script:

```sh
#!/bin/sh

echo $PATH
for x in foo fifo
do
	test -e dir/$x || continue
	/c/WINDOWS/System32/WindowsPowerShell/v1.0/powershell \
		-command 'attrib.exe -R dir\'"$x"'.lnk' || exit 1
done &&
rm -rf dir &&
mkdir dir &&
cd dir &&
set -x &&

mknod foo c 1 3 &&
ln -s foo foolnk &&
ls -l foo* >out &&
cat out &&
grep '^l.*foolnk' out || {
	echo "Failed!" >&2
	exit 1
} &&

mkfifo fifo &&
ln -s fifo fifolnk &&
ls -l fifo* >out &&
cat out &&
grep '^l.*fifolnk' out || {
	echo "Failed!" >&2
	exit 1
}
```

This works with the current MSYS2 runtime (`uname -r` reports `3.5.5-1da33849.x86_64`).

And indeed, with the `isondisk()` method it does _fail_. And with `isspecial()` it _succeeds_. I will therefore go for the latter.

Reply:
@jeremyd2019 thank you so much for your outstanding help. While looking at the code was far from pleasant, I enjoyed this collaboration more!

I reduced the fixups to a single one, as promised: a2cfafe8736436691b1fbe1bf2ea08812e78fa3e

Reply:
And of course I managed to invert the logic again 🤦

Reply:
Here is what I hope will be the tip of the final revision before we can go ahead and merge: 16f69cdec22aad0ddefaf3bb30dcebfe34175f62

Reply:
And [here](https://github.com/git-for-windows/msys2-runtime/actions/runs/12997483342?pr=84) is the corresponding Git for Windows PR build that will also exercise Git's test suite.

Reply:
*slight* editorial on commit message
> // otherwise fall back to creating a `.lnk` file

that fall back is the sysfile link type, not the `.lnk` link type.

Reply:
Fair! Feel free to make the change, otherwise I'll do it tomorrow.

Reply:
Another point in favor of `!isspecial()` being the proper condition:
https://github.com/msys2/msys2-runtime/blob/2c0e426da70f161af3776ffaa3559498b81c9df4/winsup/cygwin/path.cc#L1941-L1945

Reply:
Excellent find!

Reply:
> ```shell
> for x in foo fifo
> do
> 	test -e dir/$x || continue
> 	/c/WINDOWS/System32/WindowsPowerShell/v1.0/powershell \
> 		-command 'attrib.exe -R dir\'"$x"'.lnk' || exit 1
> done &&
> ```

I was about to debug why removing device files was failing, but I couldn't reproduce it with 3.5.6.  It looks like Corinna fixed that with the 002aad0410183153d8411405ed38c6c54c418ef5 change to `unlink` in syscalls.cc, so this workaround shouldn't be needed anymore.  Also, there was no need to call out to powershell, I was doing

```shell
for f in deepcopy/*; do attrib -r $f.lnk; rm $f.lnk; done
```

Reply:
Right, I was testing with v3.5.5 (to verify that the expectations are correct), and at first tried a pure PowerShell solution to remove the files. It did not occur to me to skip PowerShell once it worked as desired...